components/supervisor/pkg/serverapi/publicapi.go

```diff
@@ -0,0 +1,365 @@
+	OpenPort(ctx context.Context, workspaceID string, port *gitpod.WorkspaceInstancePort) (res *gitpod.WorkspaceInstancePort, err error)
+	InstanceUpdates(ctx context.Context, instanceID string, workspaceID string) (<-chan *gitpod.WorkspaceInstance, error)
+
+	// Remove this and use segment client directly
+	TrackEvent(ctx context.Context, event *gitpod.RemoteTrackMessage) (err error)
+
+	// Metrics
+	RegisterMetrics(registry *prometheus.Registry) error
+}
+
+const (
+	// KindGitpod marks tokens that provide access to the Gitpod server API.
+	KindGitpod = "gitpod"
+)
+
+type ServiceConfig struct {
+	Host              string
+	Endpoint          string
+	InstanceID        string
+	WorkspaceID       string
+	SupervisorVersion string
+}
+
+type Service struct {
+	cfg         *ServiceConfig
+	experiments experiments.Client
+
+	token   string
+	ownerID string
+
+	lastServerInstance *gitpod.WorkspaceInstance
+
+	// gitpodService server API
+	gitpodService gitpod.APIInterface
+	// publicAPIConn public API publicAPIConn
+	publicAPIConn    *grpc.ClientConn
+	publicApiMetrics *grpc_prometheus.ClientMetrics
+
+	previousUsingPublicAPI bool
+}
+
+var _ APIInterface = (*Service)(nil)
+
+func NewServerApiService(ctx context.Context, cfg *ServiceConfig, tknsrv api.TokenServiceServer) APIInterface {
+	tknres, err := tknsrv.GetToken(context.Background(), &api.GetTokenRequest{
+		Kind: KindGitpod,
+		Host: cfg.Host,
+		Scope: []string{
+			"function:getToken",
+			"function:openPort",
+			"function:trackEvent",   // do we need this?
+			"function:getWorkspace", // do we need this?
+		},
+	})
+	if err != nil {
+		log.WithError(err).Error("cannot get token for Gitpod API")
+		return nil
+	}
+	// server api
+	gitpodService, err := gitpod.ConnectToServer(cfg.Endpoint, gitpod.ConnectToServerOpts{
+		Token: tknres.Token,
+		Log:   log.Log,
+		ExtraHeaders: map[string]string{
+			"User-Agent":              "gitpod/supervisor",
+			"X-Workspace-Instance-Id": cfg.InstanceID,
+			"X-Client-Version":        cfg.SupervisorVersion,
+		},
+	})
+	if err != nil {
+		log.WithError(err).Error("cannot connect to Gitpod API")
+		return nil
+	}
+
+	service := &Service{
+		token:         tknres.Token,
+		gitpodService: gitpodService,
+		cfg:           cfg,
+		experiments: experiments.NewCustomClient(configcat.Config{
+			SDKKey:       "gitpod",
+			BaseURL:      fmt.Sprintf("%s%s", os.Getenv("GITPOD_HOST"), "/configcat"),
+			PollInterval: time.Minute * 1,
+			HTTPTimeout:  3 * time.Second,
+		}),
+		publicApiMetrics: grpc_prometheus.NewClientMetrics(),
+	}
+
+	service.publicApiMetrics.EnableClientHandlingTimeHistogram(
+		// it should be aligned with https://github.com/gitpod-io/gitpod/blob/196a109eee50bfb7da2c6b858a3e78f2a2d0b26f/install/installer/pkg/components/ide-metrics/configmap.go#L199
+		grpc_prometheus.WithHistogramBuckets([]float64{.005, .025, .05, .1, .5, 1, 2.5, 5, 30, 60, 120, 240, 600}),
```

Comment:
Do you really need such detail on the histogram? Would it make a difference if your buckets were 0.1, 0.2, 0.5, 1, 2, 5, 10?